Update quotas content and set Manage Event Stream Guide as a subtopic (#1566)
getsentry/sentry-docs: src/collections/_documentation/accounts/quotas/index.md

---
title: 'Quotas & Events'
sidebar_order: 0
---

Events and quotas are interconnected in Sentry. At the most basic, when you [subscribe to Sentry](https://sentry.io/pricing/), you pay for the number of events to be tracked. This number of events is your quota. When an event is accepted by Sentry, it counts toward your quota.

Sentry’s flexibility means you can exercise fine-grained control over which events count toward your quota.

## Key Terms

Let’s clarify a few terms to start:

> -   Event - an event is one instance of you sending Sentry data. Generally, this data is an error. Every event has a unique set of characteristics, called its fingerprint.

Reviewer: I'm realizing that the description of the "fingerprint" here is not correct. The only unique characteristic of each event is its `event_id`. The fingerprint is not unique, in contrast, as you explain below.

Author: Thank you. I've removed "unique" here.

-   Issue - an issue is a grouping of similar events, who all share the same fingerprint. For example, Sentry groups events together when they are triggered by the same part of your code. For more information, see [Grouping & Fingerprinting](https://docs.sentry.io/data-management/event-grouping/).
-   Quota - your quota is the monthly number of events you pay Sentry to track.

## What Counts Toward My Quota, an Overview

Sentry completes a thorough evaluation of each event to determine if it counts toward your quota, as outlined in this overview. Detailed documentation for each evaluation is linked throughout. Before completing any of these evaluations, Sentry confirms that each event includes a valid DSN and project as well as whether the event can be parsed and contains valid fingerprint information; if any of these items are missing or incorrect, the event is rejected.

1. **SDK configuration**

>       The SDK configuration either allows the event or filters the event out. For more information, see [Outbound Filters in our guide to Manage Your Event Stream]({%- link _documentation/accounts/quotas/manage-event-stream-guide.md -%}#outbound-filters) or [Filtering Events]({%- link _documentation/error-reporting/configuration/index.md -%}).

Reviewer: Did you mean to link to `/error-reporting/configuration/filtering.md`?

Author: Yes. Changed and tested the link. Thank you.

> 2. **SDK sample rate**

Author: Added content clarifying that SDK sample rate is not necessarily the best choice, with a pointer to rate limits.

3. **Quota availability**

      Events that exceed your quota are not sent. To add to your quota or review what happens when you exceed it, see [Increasing Quotas]({%- link _documentation/accounts/quotas/index.md -%}#increasing-quotas).

4. **Event repetition**
  -   If you have intervened to Delete and Discard events with the same fingerprint, the event does not count toward your quota.
  -   If the previous event was resolved, this event counts toward your quota because it may represent a regression in your code.
  -   If you have intervened to ignore alerts about events with the same fingerprint, this event counts toward your quota because the event is still occurring. For more information, see [Inbound Filters]({%- link _documentation/accounts/quotas/index.md -%}#inbound-data-filters).

5. **Spike protection**

      Sentry’s spike protection prevents huge overages from consuming your event capacity. For more information, see [Spike Protection]({%- link _documentation/accounts/quotas/index.md -%}#spike-protection).

In addition, depending on your project’s configuration and the plan you subscribe to, Sentry may also check:

> 6. **Rate limit for the project**

Reviewer 2: On second thought, this page you're making is for a general overview of what counts towards the quota. My previous comment was addressing more of a BestPractices or ManageYourEventStream mindset, so it should be okay here.

I feel like a lot of people coming to this page, might be coming specifically to figure out how to reduce their volume, but as long as they read, they'll find the ManageYourEventStream guide, as well as these two sections you've nicely grouped together here - InboundFilters + Rate Limiting

> 6. **Rate limit for the project**

Author: Thanks!

{% include components/alert.html
    title="Note"
    content="Per-key rate limiting is available only for Business and Enterprise plans"
    level="warning"
%}

## Inbound Filters {#inbound-data-filters}

In some cases, the data you’re receiving in Sentry is hard to filter, or you don’t have the ability to update the SDK’s configuration to apply the filters. Sentry provides several methods to filter data server-side, which apply before checking for potential rate limits. Inbound filters include:

-     Common browser extension errors
-     Events coming from localhost
-     Known legacy browsers errors
-     Known web crawlers
-     By their error message
-     From specific release versions of your code
-     From certain IP addresses.

Explore these by navigating to **[Project] » Project Settings » Inbound Filters**. Commonly-set filters are discussed here for your quick reference. For additional information and examples, see [Inbound Data Filters in our guide to Manage Your Event Stream]({%- link _documentation/accounts/quotas/manage-event-stream-guide.md -%}#inbound-data-filters). Commonly-set filters include:

> *IP Blocklist.* If you have a rogue client, Sentry supports blocking an IP from sending data. Navigate to **[Project] » Project Settings » Inbound Filters** to add the IP addresses (or subnets) to **Filter errors from these IP addresses**.

Reviewer: ```suggestion
*IP Blacklist.* If you have a rogue client, Sentry supports blocking an IP from sending data. Navigate to **[Project] » Project Settings » Inbound Filters** to add the IP addresses (or subnets) to **Filter errors from these IP addresses**.
```

Author: Will fix. Thank you.

When spike protection is activated, we limit the number of events accepted in any minute to:

`maximum(20, 6 x average events per minute over the last 24 hours)`

_The 24 hour window ends at the beginning of the current hour, not at the current minute._

What this means is that if you experience a spike, we will temporarily protect you, but if the increase in volume is sustained, the spike protection limit will gradually increase until Sentry finally accepts all events.

### How Does Spike Protection Help? {#how-does-spike-protection-help}

Because Sentry bills on monthly event volume, spikes can consume your Sentry capacity for the rest of the month. If it really is a spike, spike protection drops events during the spike to try and conserve your capacity. We also send an email notification to the Owner when spike protection is activated.

## Attribute Limits

Sentry imposes hard limits on various components within an event. While the limits may change over time and vary among attributes, most individual attributes are capped at 512 characters. Additionally, certain attributes also limit the maximum number of items. Commonly-trimmed values might include, for example:

> -     Stack frame variables. Stack traces with large frame counts will be trimmed (the middle frames are dropped). We limit to a maximum of 250 frames. Only 50 are retained unaltered (modulo usual trimming), with the remaining 200 removing `vars`, `pre_context` and `post_context`.

Reviewer: ```suggestion
-     Stack frame details. Stack traces with large frame counts will be trimmed (the middle frames are dropped). We limit to a maximum of 250 frames. Only 50 are retained unaltered, with the remaining 200 removing `vars`, `pre_context` and `post_context`.
```

Author: Thank you.

-     `extra` data is limited to approximately 256k characters, and each item is capped at approximately 16k characters.
-     Breadcrumbs and their values

> -     Generic attributes like the event’s label also have limits, but are more flexible depending on their case. For example,  messages and message parameters are limited to 1024 characters.

Reviewer: This is a bit verbose since we're already listing examples.

```suggestion
-     Error messages and message parameters are limited to 1024 characters.
```

Author: Thank you.

---
title: 'Quotas & Events'
sidebar_order: 0
---

> Events and quotas are interconnected in Sentry. At the most basic, when you [subscribe to Sentry](https://sentry.io/pricing/), you pay for the number of events to be tracked. This number of events is your quota. When an event occurs, it counts toward your quota.

Author: I agree. Good point. I will edit to "When an event is accepted by Sentry, it counts toward your quota."

-   Quota - your quota is the monthly number of events you pay Sentry to track.

## What Counts Toward My Quota, an Overview

Sentry completes a thorough evaluation of each event to determine if it counts toward your quota, as outlined in this overview. Detailed documentation for each evaluation is linked throughout. Before completing any of these evaluations, Sentry confirms that each event includes a valid DSN and project as well as whether the event can be parsed and contains valid fingerprint information; if any of these items are missing or incorrect, the event is rejected.

1. **SDK configuration**

      The SDK configuration either allows the event or filters the event out. For more information, see [Outbound Filters in our guide to Manage Your Event Stream]({%- link _documentation/accounts/quotas/manage-event-stream-guide.md -%}#outbound-filters) or [Filtering Events]({%- link _documentation/error-reporting/configuration/index.md -%}).

2. **SDK sample rate**

>       If a sample rate is defined for the SDK, Sentry evaluates whether this event should be sent as a representative fraction of the sampled events. Setting a sample rate is documented for each SDK. For more information, see [Configuration, sampleRate]({%- link _documentation/error-reporting/configuration/index.md -%}#sample-rate).

Author: Thank you. Will clarify.

3. **Quota availability**

      Events that exceed your quota are not sent. To add to your quota or review what happens when you exceed it, see [Increasing Quotas]({%- link _documentation/accounts/quotas/index.md -%}#increasing-quotas).

4. **Event repetition**

>   -   If you have intervened to discard events with the same fingerprint, the event does not count toward your quota. For more information, see [Delete & Discard in our guide to Manage Your Event Stream]({%- link _documentation/accounts/quotas/manage-event-stream-guide.md  -%}#delete--discard) or [Inbound Filters]({%- link _documentation/accounts/quotas/index.md -%}#inbound-data-filters).

Reviewer: > to discard events with the same fingerprint

Can we clarify this by literally mentioning _"Delete and Discard"_? I only realized what this means when reading the hash tag in the link.

Also, "Inbound filters" are quite a different concept and mentioned explicitly below -- might it be confusing to mention it here?

Author: Good catch. Will edit to: If you have intervened to Delete and Discard events with the same fingerprint, the event does not count toward your quota.

Reviewer 2: @PeloWriter 
Also, you discard an Issue not an event. maybe write "if you intervened to discard an Issue, then future events with the same fingerprint will not be accepted/counted towards your quota"

Also be careful here, because a really common question is, "If I Delete&Discard an issue, does that mean I won't get billed for the 10,000 events already accepted?" The answer is, "No. those still count as they were already accepted. Everything after 10,000 is what will be rejected (and not billed for)"

so make sure we're not implying that the historical events (already accepted) are not getting billed for.

Author: Thanks. Updated to discuss *future* events more clearly here and in the table.

5. **Spike protection**

      Sentry’s spike protection prevents huge overages from consuming your event capacity. For more information, see [Spike Protection]({%- link _documentation/accounts/quotas/index.md -%}#spike-protection).

In addition, depending on your project’s configuration and the plan you subscribe to, Sentry may also check:

6. **Rate limit for the project**

      If the rate limit for the project has been exceeded, and your subscription allows, the event will not be counted. For more information, see [Rate Limiting in our guide to Manage Your Event Stream]({%- link _documentation/accounts/quotas/manage-event-stream-guide.md -%}#4-rate-limiting) or  [Rate Limiting Projects]({%- link _documentation/accounts/quotas/index.md -%}#id1).

7. **Inbound Filters**

      If any inbound filter is set for this type of event, and your subscription allows, the event will not be counted. For more information, see [Inbound Filters in our guide to Manage Your Event Stream]({%- link _documentation/accounts/quotas/manage-event-stream-guide.md -%}#inbound-data-filters) or [Inbound Data Filter]({%- link _documentation/accounts/quotas/index.md -%}#inbound-data-filters).

> After these checks are processed, the event counts toward your quota. It is accepted into Sentry, where it persists and is stored.

Reviewer: What might be worth mentioning: Events from disabled DSNs do **not** count against the quota.

Author: That is helpful. Yes, I will add. Thanks.

Author: Actually, we have it already at line 19 - "Before completing any of these evaluations, Sentry confirms that each event includes a valid DSN and project as well as whether the event can be parsed and contains valid fingerprint information; if any of these items are missing or incorrect, the event is rejected." 

I don't think it needs repeating?

Reviewer: You’re right. If people jump right to this list,they might maybe miss it, but since “disabling” a DSN is quite straight-forward, it’s not worth repeating here.

A few important notes:

>   - For mobile users, a storage fee is also charged if the Native SDK minidump is attached to the event in addition to the event counting toward your quota. For more information, see [Event Attachments (Preview)]({%- link _documentation/platforms/native/minidump.md -%}#event-attachments-preview).

Reviewer: This is not a mobile use case. You can just mention: 

> A storage fee is charged for minidumps and event attachments on stored events. For more information, ...

Also, note that this is not GA yet and subject to special sales agreements, until we roll out the extended quotas for attachment storage (after which the above definition of quota no longer holds, as it also applies to non-event items).

Author: I will make the edit to clarify it's true for all, not just mobile. But since it is not yet GA, would you like this commented out for now?

Reviewer: Commenting it out would be better for now, yes.

Author: Done. Thanks.

>   - If the event exceeds 200KB compressed for events and 20MB compressed for minidump uploads (all files combined), the event will be rejected. For more information, see [Attribute Limits]({%- link _documentation/accounts/quotas/index.md -%}#attribute-limits).

Reviewer: We **also** enforce 1MB decompressed for events and 50MB decompressed for minidump uploads.

Author: Added. Thanks.

To set up rate limits, navigate to **[Project] » Client Keys » Configure**. Select an individual key or create a new one, then you’ll be able to define a rate limit as well as view a breakdown of events received by that key. For additional information and examples, see [Rate Limiting in our guide to Manage Your Event Stream]({%- link _documentation/accounts/quotas/manage-event-stream-guide.md -%}#4-rate-limiting).

{% include components/alert.html
    title="Note"
    content="Per-key rate limiting is available only for Business and Enterprise plans"
    level="warning"
%}

## Inbound Filters {#inbound-data-filters}

In some cases, the data you’re receiving in Sentry is hard to filter, or you don’t have the ability to update the SDK’s configuration to apply the filters. Sentry provides several methods to filter data server-side, which apply before checking for potential rate limits. Inbound filters include:

-     Common browser extension errors
-     Events coming from localhost
-     Known legacy browsers errors

> -     Custom filters to filter out errors

Reviewer: Which filter does this map to?

Author: Line 154 is old content. I will remove it.

-     By their error message
-     From specific release versions of your code

> -     From certain IP addresses.

Reviewer: Web crawlers could be added to this list.

Author: Will add. Thank you.

### How Does Spike Protection Help? {#how-does-spike-protection-help}

Because Sentry bills on monthly event volume, spikes can consume your Sentry capacity for the rest of the month. If it really is a spike, spike protection drops events during the spike to try and conserve your capacity. We also send an email notification to the Owner when spike protection is activated.

## Attribute Limits

Sentry imposes hard limits on various components within an event. While the limits may change over time and vary among attributes, most individual attributes are capped at 512 characters. Additionally, certain attributes also limit the maximum number of items.

For example, `extra` data is limited to approximately 256k characters, and each item is capped at approximately 16k characters.

Generic attributes like the event’s label also have limits, but are more flexible depending on their case. For example, the message attribute is limited to 1024 characters.

The following limitations are automatically enforced:

-   Events greater than 200KB are immediately dropped (pre decompression).

> -   Stack traces with large frame counts will be trimmed (the middle frames are dropped).

Reviewer: We limit to a maximum of `250` frames. Only `50` are retained unaltered (modulo usual trimming), with the remaining `200` removing `vars`, `pre_context` and `post_context`.

Author: Added. Thank you.
